File: 1249-minimum-remove-to-make-valid-parentheses/1249. Minimum Remove to Make Valid Parentheses.py
class Solution:
    def minRemoveToMakeValid(self, s: str) -> str:
        res = ["1"] * len(s)
        stack = []

        for i in range(len(s)):
            if s[i] == ")":
                if not stack:
                    continue
                idx = stack.pop()
                res[idx] = "("
                res[i] = ")"
            elif s[i] == "(":
                stack.append(i)
            else:
                res[i] = s[i]
        print(res)
        final_res = ""
        for c in res:
            if c != "1":
                final_res += c
        return final_res

File: 1249-minimum-remove-to-make-valid-parentheses/1249. Minimum Remove to Make Valid Parentheses.cpp
class Solution {
public:
    string minRemoveToMakeValid(string s) {
        vector<char> res(s.size(), '\0');
        stack<int> stack;

        for (int i = 0; i < s.length(); i++) {
            if (s[i] == ')') {
                if (stack.empty()) continue;
                int idx = stack.top();
                stack.pop();
                res[idx] = '(';
                res[i] = ')';
            } else if (s[i] == '(') {
                stack.push(i);
            } else {
                res[i] = s[i];
            }
        }
        string final_res = "";
        for (int i = 0; i < res.size(); i++) {
            if (res[i]) {
                final_res += res[i];
            }
        }
        return final_res;
    }
};